SI-8521 ScalaClassLoader.asContext saves context

som-snytt opened the pull request:
It saves the actual current thread context loader,
not the loader it might biasedly prefer to have.

REPL uses asContext to preserve context, which broke
if the current loader didn't happen to be a
ScalaClassLoader.

som-snytt committed:
SI-8521 ScalaClassLoader.asContext saves context

It saves the actual current thread context loader,
not the loader it might biasedly prefer to have.

REPL uses asContext to preserve context, which broke
if the current loader didn't happen to be a
ScalaClassLoader.

som-snytt commented:
/rebuild

Internal error when running tests: sbt.ForkMain$Run$RunAborted: java.net.SocketException: Broken pipe (Write failed)

retronym reviewed (approved):
LGTM

retronym reviewed (dismissed):
Oh, wait, do we have the same problem in:

```scala
object ScalaClassLoader {
  def savingContextLoader[T](body: => T): T = {
    val saved = contextLoader
    try body
    finally setContext(saved)
  }
```

som-snytt commented:
I don't know why life can be so evil, but I'll give that API a haircut.

som-snytt committed:
SI-8521 No blind save of context class loader

The REPL would presumptively preserve the context
class loader, on the presumption that it would get
set and not restored. Since that behavior was
ameliorated, and since the method was also flawed,
just remove that altogether.

som-snytt commented:
Rebased, and I think I removed the offending method, which was a stop-gap. Since REPL now restores context class loader in a principled way, the stop-gap is not required.

I don't really know what "principled" means, except that retronym uses that word approvingly, so I hope it helps my case. It might mean, We promise to restore the context class loader. Right brace means "rewind" as in "Please be kind, rewind" from the days of VHS.

lrytz commented on src/repl/scala/tools/nsc/interpreter/IMain.scala:
```diff
@@ -359,8 +359,7 @@ class IMain(initialSettings: Settings, protected val out: JPrintWriter) extends
 
-  // Set the current Java "context" class loader to this interpreter's class loader
-  def setContextClassLoader() = classLoader.setAsContext()
+  def setContextClassLoader() = ()
```
mark `@deprecated`?

som-snytt replied:
I went too far; I'll back off the changes. Comment on another PR says sbt uses this. (Or maybe that's OK.)

lrytz reviewed (commented):
> REPL now restores context class loader in a principled way

Can you provide a reference?

som-snytt commented:
Saving context class loader was part of the original contribution for scripting:

https://github.com/scala/scala/blob/v2.12.1/src/repl/scala/tools/nsc/interpreter/IMain.scala#L563

I found the previous topic:

https://github.com/scala/scala/pull/4841

I'll refresh my memory and also maybe take up that old PR.

som-snytt commented:
As seen in https://github.com/scala/scala/pull/5655 sbt runs `initialCommands` before `process` so prophylactic `savingContextClassLoader` doesn't provide extra security.

I'm not sure if leaving the API in `object ScalaClassLoader` is nice to do, even though it is nominally internal API.

som-snytt committed:
SI-4331 Fix memory leaks across REPL :reset command

Retronym's comment to his fix:

When replacing the classloader, which retains all values created
in the prior REPL lines, we must be careful not to keep around
references to the previous classloader.

We were doing this in two places: the thread context classloader and
in an runtime reflection mirror.

Both of those references to the old classloader would have been
correctness problems, and would likely have meant some degraded
functionality for a post :reset REPL. Furthermore, they served to
retain memory allocated up to the first :reset forever.

I changed handling of the thread context classlaoder to only set and
restore it around execution of code. This avoids leaking the
classloader to whatever thread happend to initialize the REPL.
We actually call the initialization code from within in a background
thread that allows the user to start typing the first command sooner.

I trawled the git logs to understand the history of how and when
we set the context classloader. It was originally added in
7e617ef. A more recent change to
add support for JSR-223 (Scripting) also wrapped this around the
execution 3a30af1. I believe the second commit is the correct
approach, and that it safe to remove the old code.

While this commit makes :reset perform its advertised function,
it doesn't go further to making it a particulary useful command.
Ideally we'd offer a mechanism to transport some data across a :reset
boundary. The user could do something like this manually by stashing
data in a `java.util.Map` hosted in some class in the same classloader
as IMain. They would also have to restrict this to sending "pure data"
rather than instances of data types defined in the REPL session
itself.

I located the leaks with YourKit, and have tested manually by
repeatedly running:

```
 qscala -J-XX:+HeapDumpOnOutOfMemoryError -J-Xmx512M
Welcome to Scala 2.12.0-20151110-111920-44ae563d37 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_51).
Type in expressions for evaluation. Or try :help.

scala> {val b = new Array[Byte](256 * 1024* 1024); () => b}
res0: () => Array[Byte] = $$Lambda$1781/1479140596@4218500f

scala> :reset
Resetting interpreter state.
Forgetting this session history:

{val b = new Array[Byte](256 * 1024* 1024); () => b}

Forgetting all expression results and named terms: $intp

scala> {val b = new Array[Byte](256 * 1024* 1024); () => b}
res0: () => Array[Byte] = $$Lambda$1828/1370495328@6831d8fd

scala> :reset
Resetting interpreter state.
Forgetting this session history:

{val b = new Array[Byte](256 * 1024* 1024); () => b}

scala> {val b = new Array[Byte](256 * 1024* 1024); () => b}
res0: () => Array[Byte] = $$Lambda$1829/79706679@2e140e59

scala> :reset
Resetting interpreter state.
Forgetting this session history:

{val b = new Array[Byte](256 * 1024* 1024); () => b}

scala> {val b = new Array[Byte](256 * 1024* 1024); () => b}
res0: () => Array[Byte] = $$Lambda$1830/1649884294@2418ba04

scala> {val b = new Array[Byte](256 * 1024* 1024); () => b}
java.lang.OutOfMemoryError: Java heap space
Dumping heap to java_pid21760.hprof ...
Heap dump file created [346682068 bytes in 0.870 secs]
java.lang.OutOfMemoryError: Java heap space
  ... 32 elided

scala> :reset
Resetting interpreter state.
Forgetting this session history:

{val b = new Array[Byte](256 * 1024* 1024); () => b}

scala> {val b = new Array[Byte](256 * 1024* 1024); () => b}
res0: () => Array[Byte] = $$Lambda$1839/174999703@32f5ecc4
```

som-snytt commented:
Just tacking on retronym's old fix, cherry not picked. I haven't tested yet.

SethTisue commented:
@som-snytt @cnestor you pinged me on Gitter about this. yes, let's make sure this makes 2.12.4. Som, can you rebase, and comment on whether this is otherwise ready for merge...?

som-snytt commented:
:notes: When you wish upon a star... :musical_note: :zzz:

som-snytt commented:
@SethTisue Thanks, I pushed something. As summary, `IMain` uses `asContext` to save/restore context class loader, and this is deemed correct. After fixing `asContext` to work correctly, remove the similar utility method and don't use it from `ILoop.process`.

som-snytt commented:
/rebuild

som-snytt commented:
The failure mode is
```
Error: publish-core did not publish "jenkins.properties" artifact, unable to proceed.
ERROR: Failed to run DSL Script
groovy.lang.MissingPropertyException: No such property: Result for class: Script1
```

som-snytt commented:
/rebuild

lrytz commented:
```
[error] /home/jenkins/workspace/scala-2.12.x-validate-publish-core/src/repl/scala/tools/nsc/interpreter/IMain.scala:1041: stable identifier required, but IMain.this.runtimeMirror.universe found.
[error]       import runtimeMirror.universe.{Symbol, InstanceMirror, TermName}
[error]                            ^
```

som-snytt commented:
I swear I did a `testOnly ScriptedTest` yesterday, but I just realized I was probably in the middle of the rebase and the last commit hadn't been applied yet. I only had a few minutes to spare; just to confirm, testOnly does actually compile repl. I think I need a tool to tell me if I'm pushing without at least compiling and running tests that are in the commit. In fact, to run those tests because I don't want to look them up.

som-snytt commented:
@retronym 's requested change is in. I think I'm supposed to "dismiss review" but the ball is in his court?

adriaanm commented:
Let's see what this dismiss thing can do for our workflow.

adriaanm commented:
I rerequested a rereview.

adriaanm commented:
This won't merge forward to 2.13.x -- will need a separate port if we need these changes there.
